### PR TITLE
IPv6 link local support

### PR DESCRIPTION
**Pull request by cspiel1**
This PR adds IPv6 link local addresses as SIP transport.

This adds support for outgoing calls to IPv6 link local addresses and answering incoming calls from IPv6 link local peers.

### CODE REVIEW

File: src/net.c [hunk 1]
```diff
@@ -224,6 +225,20 @@ static void ipchange_handler(void *arg)
 }
 
 
+static void set_default_ipv6ll(struct network *net)
+{
+	char ifname[32] = "";
```
Comment (sreimers) [inline src/net.c:230]: `ifname[32]` is a bit small, would prefer `ifname[64]`, on windows i see interface names like this:
`{B3B5F4C1-231D-4878-8882-2F81737EEA97}`

**Comment by alfredh** [comment]
ipv6 linklocal address is also an IPv6 address, so I dont see the reason to add a new field here.

instead of storing each address for each address family, I think we should
change it into an array which makes the code more flexible.

`struct network`:

Before:

```c
	struct sa laddr;
#ifdef HAVE_INET6
	struct sa laddr6;
#endif
```

After:

```c
#define MAX_ADDR 8
	struct sa laddrv[MAX_ADDR];
```

this would also fit into the plan of supporting multiple IPv6 addresses.

**Comment by cspiel1** [comment]
From baresip point of view this should be possible.

I have to check if the libre PR https://github.com/baresip/re/pull/106 is also possible with multiple IPv6 addresses.

The crucial point will be how to implement function
```
static const struct sip_transport *transp_find(struct sip *sip,
					       enum sip_transp tp,							       enum sip_transp tp,
					       int af, const struct sa *dst)							       int af, const struct sa *dst)
```
It is called from `sip_transp_laddr()`, which is used to get the local address to copy the IPv6 scope ID to the `dst` address on different locations where a destination address could be an IPv6 link local address. Otherwise the IPv6 packet can't be routed correctly.

**Comment by alfredh** [comment]
keep in mind that net.c and ua.c are two different subsystems.

net.c is for the whole application (incl SIP,RTP,ICE etc)

ua.c is for SIP only

so we can still add support for multiple addresses in net.c

**Comment by cspiel1** [comment]
- Now net.c adds all local addresses (except loopback) to a list.
- Instead of net_laddr_af() new function net_laddr_dst() should be used wherever it is possible.
- This version allows any IPv4, IPv6, IPv6ll address to be used for peer-to-peer calls. It should not make a difference if there is a route to the internet.

Future work:
- Without a change in the API (`ua_call_alloc()`) the destination address is not available where `net_laddr_af()` is called. :-(
- Is net.c already obsolete? The information could be retrieved from the system with libre-s net API on demand.

**Comment by cspiel1** [comment]
@sreimers The build fails with the re version check.
```
ld: malformed 32-bit x.y.z version number: 1.0.0-dev
230
clang: error: linker command failed with exit code 1 (use -v to see invocation)
231
make: *** [librem.dylib] Error 1
232
Error: Process completed with exit code 2.
```

**Comment by sreimers** [comment]
I think you need to rebase the libre PR (it uses the old mk/re.mk).

**Comment by juha-h** [comment]
Does this work now solve the issue https://github.com/juha-h/baresip-studio/issues/54 that my baresip Android app has had from the very beginning, i.e., that baresip can use only one IP address for outbound requests?  For example, if the device has mobile network IP address, WiFi IP address and a VPN IP address, then baresip will choose the outbound source IP address based on OS routing table?

**Comment by cspiel1** [comment]
Yes, I think so because the new function `net_check_conn()` in the libre PR should test the connection with a temporary UDP socket.

Have to check/fix that network changes are detected and applied to active calls and the net object. Then will activate this PR again.

**Comment by cspiel1** [comment]
Now changes in the `laddrs` list work. The PR is ready for your tests and review.

As I already wrote above in the future TODOs:
- The use of `net_laddr_af()` in ua.c should be replaced by `net_laddr_dst()` if the call destination is an IP-address. But we have no destination address in `ua_call_alloc()`.
- The result is that if there are multiple interfaces the IP address in SDP might be a different than for SIP. 
- Also it is possible that e.g. SIP uses IPv6 and RTP IPv4. I know that this is normally ok. But maybe can lead to troubles.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Yes, I think so because the new function `net_check_conn()` in the
> libre PR should test the connection with a temporary UDP socket.

It is not possible to test SIP connectivity with UDP socket, since
firewall somewhere between client and server could block UDP SIP
traffic.

**Comment by sreimers** [comment]
> It is not possible to test SIP connectivity with UDP socket,

There is no real UDP Datagram send, only local udp socket is connected for test.

From `man connect`
```
    If the initiating socket is not connection-mode, then connect() shall set the socket's peer address, and no connection is made.  For  SOCK_DGRAM  sockets,  the
       peer  address  identifies  where  all  datagrams  are sent on subsequent send() functions, and limits the remote sender for subsequent recv() functions. If the
       sa_family member of address is AF_UNSPEC, the socket's peer address shall be reset. Note that despite no connection being made, the term ``connected'' is  used
       to describe a connectionless-mode socket for which a peer address has been set.
```

**Comment by cspiel1** [comment]
> > It is not possible to test SIP connectivity with UDP socket,
> 
> There is no real UDP Datagram send, only local udp socket is connected for test.
> 
> From `man connect`
> 
> ```
>     If the initiating socket is not connection-mode, then connect() shall set the socket's peer address, and no connection is made.  For  SOCK_DGRAM  sockets,  the
>        peer  address  identifies  where  all  datagrams  are sent on subsequent send() functions, and limits the remote sender for subsequent recv() functions. If the
>        sa_family member of address is AF_UNSPEC, the socket's peer address shall be reset. Note that despite no connection being made, the term ``connected'' is  used
>        to describe a connectionless-mode socket for which a peer address has been set.
> ```

For `net_check_conn()` I took the code from `net_dst_source_addr_get()`. Maybe I made a to strong statement, that the connection is tested. Only the route is checked what should be enough to select the correct local IP address, right? Anyway I will have to write a unit-test for `net_check_conn()`. Then we will see.

**Comment by sreimers** [comment]
> Only the route is checked what should be enough to select the correct local IP address, right?

I think so.

**Comment by juha-h** [comment]
I tested baresip built with pull requests:

https://github.com/baresip/re/pull/106
https://github.com/baresip/baresip/pull/1473.

on host with the following following interfaces:

$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:a9:81:ac brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
       valid_lft 84989sec preferred_lft 84989sec
    inet6 fe80::a00:27ff:fea9:81ac/64 scope link 
       valid_lft forever preferred_lft forever
3: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:b5:50:84 brd ff:ff:ff:ff:ff:ff
    inet 192.26.134.12/24 brd 192.26.134.255 scope global enp0s8
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:feb5:5084/64 scope link 
       valid_lft forever preferred_lft forever

and the following routing table:

$ route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.0.2.2        0.0.0.0         UG    0      0        0 enp0s3
10.0.2.0        0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
169.254.0.0     0.0.0.0         255.255.0.0     U     1000   0        0 enp0s3
192.26.134.0    0.0.0.0         255.255.255.0   U     0      0        0 enp0s8
192.168.43.0    192.26.134.1    255.255.255.0   UG    1      0        0 enp0s8

In the first test, destination was ***@***.*** where tutpro.com is
reachable via default route gateway 10.0.2.2.  In that test INVITE was
sent like this:

TLS 10.0.2.15:40128 -> 192.26.111.38:8003

and the local IP address in INVITE headers and SDP correctly was
10.0.2.15.

Then I made another test where destination was ***@***.*** where
test.tutpro.com IP address 192.168.43.160 is reachable via gateway
192.26.134.1.  In that test the local IP address in INVITE headers and
SDP was again 10.0.2.15:

TLS 192.26.134.12:34080 -> 192.168.43.160:5061
INVITE ***@***.*** SIP/2.0
Via: SIP/2.0/TLS 10.0.2.15:5061;branch=z9hG4bKc737df1a879c24cf;rport
Contact: ***@***.***:5061;transport=tls>
Max-Forwards: 70
To: ***@***.***>
From: ***@***.***>;tag=9806c1a997cebd11
Call-ID: 3eac1754e98c4821
CSeq: 48392 INVITE
User-Agent: baresip v1.1.0 (x86_64/linux)
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,REFER
Supported: gruu
Content-Type: application/sdp
Content-Length: 320

I think that is incorrect.  IP address in INVITE headers and SDP should
have been 192.26.134.12.

**Comment by juha-h** [comment]
Looks like github censored my comment.  Let me know if you need the censored bits.

**Comment by cspiel1** [comment]
Removed `net_check_conn()` in libre. It didn't check the route correctly. Instead use `net_dst_source_addr_get()` which provides a useful local address for a given destination.

@juha-h Thanks for the test. Please could you try again with current branches ipv6ll_support? I think, you are right with what you are expecting for your setup.

**Comment by juha-h** [comment]
So it does not make sense for me to do another test?

I tried, but baresip didn't compile using latest pull request patched re and baresip:
```
 CC      build-x86_64/src/net.o
src/net.c: In function ‘find_laddr_af’:
src/net.c:656:32: warning: implicit declaration of function ‘net_check_conn’; did you mean ‘net_check’? [-Wimplicit-function-declaration]
   if ((lc & LADDR_INTERNET) && net_check_conn(&laddr->sa, &dst))
                                ^~~~~~~~~~~~~~
                                net_check
src/net.c:656:32: warning: nested extern declaration of ‘net_check_conn’ [-Wnested-externs]
```

**Comment by cspiel1** [comment]
Maybe a new field in struct ua used temporary.
```
struct ua {
...
	struct sa dst;               /**< Current destination address        */
};
```

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Maybe a new field in struct ua used temporary.
> ```
> struct ua {
> ...
> 	struct sa dst;               /**< Current destination address        */
> };
> ```

How could ua have a current destination address, since I can have an
ongoing call from the ua with somebody and at the same time send
messages from the ua to other parties?

**Comment by cspiel1** [comment]
`ua->dst` is set before `ua_call_alloc()` is called and directly used in it, then cleared immediately. Of course it is more convenient to add it to the parameter list of `ua_call_alloc()`. Should we?

**Comment by sreimers** [comment]
```
[ RUN      ] test_call_answer
ua: SDP offer origin address 127.0.0.1 not reachable
selftest: re_main() loop timed out -- test hung..
TEST_ERR: test/call.c:364: (Operation timed out)
test_call_answer: test failed (Operation timed out)
test failed (Operation timed out)
```
Why test_call_answer with `127.0.0.1` fails on macos is a little strange (because `net_dst_source_addr_get` with 127.0.0.1 works in retest on macos). I can investigate this or is this related to `net_laddr_af` ?

**Comment by juha-h** [comment]
Christian Spielberger writes:

> `ua->dst` is set before `ua_call_alloc()` is called and directly used in it, then cleared immediately. Of course it is more convenient to add it to the parameter list of `ua_call_alloc()`. Should we?

Is ua_call_alloc used also when MESSAGE, OPTIONS, and PUBLISH requests
are sent?

**Comment by juha-h** [comment]
Kamailio SIP Proxy has this file that includes socket related stuff:

  https://github.com/kamailio/kamailio/blob/master/src/core/forward.c

For example, it has function get_out_socket() that

* return a socket_info_pointer to the sending socket; as opposed to
 * get_send_socket, which returns process's default socket, get_out_socket
 * attempts to determine the outbound interface which will be used;
 * it uses a temporary connected socket to determine it; it will
 * be very likely noticeably slower, but it can deal better with
 * multihomed hosts
 */

Then it has forward_request() function that calls

  build_req_buf_from_sip_req(msg, &len, send_info, 0);

which rebuilds the SIP message based on the sending socket.

Makes sense to me to handle the source IP/port stuff at low level where
there is no knowledge left about UAs and other high level SIP concepts.

**Comment by cspiel1** [comment]
Now it works also "correct" for outgoing calls. That means if the destination is an IP-address:
- The address family (af) of the destination is used for RTP.
- The local IP address is chosen for SDP that has a route to the destination IP.
- If destination is a IPv6ll it simply works. Also in SDP we have the local IPv6ll address of the correct interface.

Your questions:
- @juha-h MESSAGE, OPTIONS and PUBLISH seems to have not a real call. Only handling of OPTIONS has a temp dummy call to encode the SDP answer. See `ua_handle_options()`! We should pass the `sip_msg` to `ua_call_alloc` and set af and laddr adequate.
- @juha-h The local address is needed in the `sdp_session`. Currently it is collected in baresip-s net object and passed from ua to call to the `sdp_session`. Maybe we could get rid of collecting laddrs in the net object and retrieve the right local address on the fly in the SDP is encoded. I am not sure if it is possible.
- @sreimers I have no idea about the problem on MAC-OS or the error "Can't assign requested address". Could you please debug this?

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Now it works also "correct" for outgoing calls. That means if the
> destination is an IP-address:

In my tests, destination was not an IP address, but a domain name that is
resolved via naptr/srv/a process.

Why not just mark in the message where the source address need to be
placed and let the low level part of the code replace the mark with real
address once the destination has been resolved and socket chosen?

That would be independent of what kind of message is being sent out.

There could also exist a "force send socket" account parameter if user
wants to override the routing table one.

**Comment by sreimers** [comment]
> * @sreimers I have no idea about the problem on MAC-OS or the error "Can't assign requested address". Could you please debug this?

The destination port must not be zero, i just tested successfully this patch:

```diff
diff --git a/src/net/net.c b/src/net/net.c
index 70f7b25b..5cf928d5 100644
--- a/src/net/net.c
+++ b/src/net/net.c
@@ -70,6 +70,7 @@ int net_dst_source_addr_get(const struct sa *dst, struct sa *ip)
 {
        int err;
        struct udp_sock *us;
+       struct sa tmp_dst;
 
        if (sa_af(dst) == AF_INET6)
                err = sa_set_str(ip, "::", 0);
@@ -79,11 +80,15 @@ int net_dst_source_addr_get(const struct sa *dst, struct sa *ip)
        if (err)
                return err;
 
+       tmp_dst = *dst;
+       if (sa_port(&tmp_dst) == 0)
+               sa_set_port(&tmp_dst, 53);
+
        err = udp_listen(&us, ip, NULL, NULL);
        if (err)
                return err;
 
-       err = udp_connect(us, dst);
+       err = udp_connect(us, &tmp_dst);
        if (err)
                goto out;
```

Would you like to include this on your PR or should I make a separate PR and you rebase?

**Comment by sreimers** [comment]
Maybe we simply fallback to `net_default_source_addr_get` if no destination address is available?

**Comment by cspiel1** [comment]
> > * @sreimers I have no idea about the problem on MAC-OS or the error "Can't assign requested address". Could you please debug this?
> 
> The destination port must not be zero, i just tested successfully this patch:
> 
> ```diff
> diff --git a/src/net/net.c b/src/net/net.c
> index 70f7b25b..5cf928d5 100644
> --- a/src/net/net.c
> +++ b/src/net/net.c
> @@ -70,6 +70,7 @@ int net_dst_source_addr_get(const struct sa *dst, struct sa *ip)
>  {
>         int err;
>         struct udp_sock *us;
> +       struct sa tmp_dst;
>  
>         if (sa_af(dst) == AF_INET6)
>                 err = sa_set_str(ip, "::", 0);
> @@ -79,11 +80,15 @@ int net_dst_source_addr_get(const struct sa *dst, struct sa *ip)
>         if (err)
>                 return err;
>  
> +       tmp_dst = *dst;
> +       if (sa_port(&tmp_dst) == 0)
> +               sa_set_port(&tmp_dst, 53);
> +
>         err = udp_listen(&us, ip, NULL, NULL);
>         if (err)
>                 return err;
>  
> -       err = udp_connect(us, dst);
> +       err = udp_connect(us, &tmp_dst);
>         if (err)
>                 goto out;
> ```
> 
> Would you like to include this on your PR or should I make a separate PR and you rebase?

Thanks! I will include this.

**Comment by cspiel1** [comment]
> In my tests, destination was not an IP address, but a domain name that is resolved via naptr/srv/a process. Why not just mark in the message where the source address need to be placed and let the low level part of the code replace the mark with real address once the destination has been resolved and socket chosen? That would be independent of what kind of message is being sent out. There could also exist a "force send socket" account parameter if user wants to override the routing table one.

I will have a look on it. One complication what I know currently is that the audio/video objects and their stream objects are allocated very early. The stream objects are opening their sockets. For this the address family has to be known.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> I will have a look on it. One complication what I know currently is
> that the audio/video objects and their stream objects are allocated
> very early. The stream objects are opening their sockets. For this the
> address family has to be known.

Address family is not known before dns lookup and the lookup may result
in both ipv4 and ipv6 addresses and more than one of those may need to
be tried before the server is reached.  So the both SIP headers and SDP
may need to be changed.

Perhaps the stream object has to have place holders for address and
address family related fields.

**Comment by cspiel1** [comment]
@juha-h 
- It is possible to postpone the choice of laddr directly before the call of `call_sdp_get()` where the SDP is encoded.
- Then we have also to solve the complication with the address family for the RTP sockets. If the `af` does not match you get:
```
stream: rtp_send to open natpinholefailed (Bad file descriptor)
video: video stream error: Bad file descriptor
```
- Moving the choice to an even lower layer is not possible, because the target address for SIP is not available in the `sdp_session` or the `rtp` part. So we have to solve this in baresip call.c and ua.c. Directly before the calls to `call_sdp_get()`. Or in `call_sdp_get()` if we add a parameter `struct sa *dst`.

**Comment by cspiel1** [comment]
> Christian Spielberger writes:
> I will have a look on it. One complication what I know currently is that the audio/video objects and their stream objects are allocated very early. The stream objects are opening their sockets. For this the address family has to be known.
> Address family is not known before dns lookup and the lookup may result in both ipv4 and ipv6 addresses and more than one of those may need to be tried before the server is reached. So the both SIP headers and SDP may need to be changed. Perhaps the stream object has to have place holders for address and address family related fields.

Ok, maybe possible if the socket open can be postponed also. I will check this.

**Comment by cspiel1** [comment]
> * The result of the DSN lookup should be available already.

The DNS lookup result is not available when the SDP is encoded. I mixed up DSN lookup with NAT resolution. So the SDP encoding would also have to be postponed.

**Comment by cspiel1** [comment]
The current solution works if the destination address is already an IP-address. This is what we need.

An improvement for domain addresses seems to be not possible without changing the API in re/baresip.

I will solve the failed check now. And try to write a retest for more test coverage.

**Comment by alfredh** [comment]
we should also discuss if `src/net.c` is obsolete.

currently it is a thin wrapper on top of `re/src/net`.

the main functions are:

- storing of local IP address(es)
- storing of DNS client
- timer to detect network changes

**Comment by cspiel1** [comment]
> we should also discuss if `src/net.c` is obsolete.
> 
> currently it is a thin wrapper on top of `re/src/net`.
> 
> the main functions are:
> 
> * storing of local IP address(es)
> * storing of DNS client
> * timer to detect network changes

I agree, thanks for the list.
- local IP addresses. --> They can be retrieved on demand from the system.
- DNS clients. --> The same?
- A timer for local IP changes is only needed if there is currently a call that should be updated like it is done with `call_reset_transp()`.

**Comment by cspiel1** [comment]
In order to detect any change in the list of local addresses, we need to store the list to compare it periodically. For this purpose it makes sense to have it stored in net.c.

Instead we can compare the systems local addresses with the laddr of the transport objects in `uag.sip`.  Right?

File: src/call.c [hunk 2]
```diff
@@ -2104,6 +2104,19 @@ struct audio *call_audio(const struct call *call)
...
+/**
+ * Get the remote IP address for the given call
+ *
+ * @param call  Call object
+ *
+ * @return Remote IP address
+ */
+const struct sa *call_raddr(const struct call *call)
+{
+	return stream_raddr(audio_strm(call_audio(call)));
```
Comment (alfredh) [inline src/call.c:2116]: what about video stream ?

File: include/baresip.h [hunk 3]
```diff
@@ -689,6 +690,7 @@ const char *menc_event_name(enum menc_event event);
 struct network;
 
 typedef void (net_change_h)(void *arg);
+typedef int (net_laddr_h)(const struct sa *sa);
```
Comment (alfredh) [inline include/baresip.h:693]: perhaps also add interface name ?

File: src/call.c (hunk 2 above)
Comment (cspiel1) [reply src/call.c:2116]: If there is a `c` line in the remote SDP then it is possible that raddr differ for audio and video. But both should be reachable from same local network interface and local address. The result for
```
net_dst_source_addr_get(call_raddr(call),
							     &laddr)
```
should be the same.

The name of the function might be confusing. But we don't need it. I will drop it and use the underlying functions directly.

File: include/baresip.h (hunk 3 above)
Comment (cspiel1) [reply include/baresip.h:692]: Yes, this makes sense.

**Comment by cspiel1** [comment]
Removal of baresip/src/net.c:
- Removal of the laddr list seems to be possible and makes sense to reduce redundancy in data.
- Removal of nsv, nsvf. This is also redundancy. It is duplicated from the config.
- Removal of dnsc. I am not sure if this would improve something. It makes sense that the DNS clients are allocated only once and re-used for each DNS lookup. Further we should not move this all to uag.c. Right?
- Check of network change could also stay inside net.c, if it is able to compare with the laddrs of the transp list.

**Comment by sreimers** [comment]
> * Removal of nsv, nsvf. This is also redundancy. It is duplicated from the config.

Is not redundant because its a special use-case for nameserver fallbacks. I have many customers with defect nameservers, but I don't want enforce for privacy something like 8.8.8.8 for all customers (in some restricted networks they also not work, so overriding is not an option). If no `dns_server` is configured, baresip detects the system ones and with `dns_fallback` I can define additional ones to query (especially SRV querys are blocked sometimes). `dns_server` is used to ignore the system ones.

File: include/baresip.h (hunk 3 above)
Comment (cspiel1) [reply include/baresip.h:692]: But if we drop the net->laddrs list, then this function pointer is obsolete.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> The current solution works if the destination address is already an
> IP-address. This is what we need.

I don't think it is a good idea to introduce solutions that work only in
some not so common special cases.

> An improvement for domain addresses seems to be not possible without
> changing the API in re/baresip.

It would be better to fix the API in re/baresip and have a solution that
works also in the very common case where server address is domain name.

**Comment by juha-h** [comment]
Sebastian Reimers writes:

> Is not redundant because its a special use-case for nameserver
> fallbacks. I have many customers with defect nameservers, but I don't
> want enforce for privacy something like 8.8.8.8 for all customers (in
> some restricted networks they also not work, so overriding is not an
> option). If no `dns_server` is configured, baresip detects the system
> ones and with `dns_fallback` I can define additional ones to query
> (especially SRV querys are blocked sometimes). `dns_server` is used to
> ignore the system ones.

Just to note that Android does not allow applications to access system
name server list.

**Comment by alfredh** [comment]
I would also like to suggest to split up this patch into logical parts.


for SIP multi-homing it should be quite easy to do it for TCP/TLS:

1. TCP connect to the SIP server
2. when connected, look at IP-address in `tcp_conn_local_get`
3. use this address in SIP via header and other headers for local IP

this should work if the device has multiple network interfaces.

The SDP part is not so easy, and should be handled separately.

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> for SIP multi-homing it should be quite easy to do it for TCP/TLS:
> 
> 1. TCP connect to the SIP server
> 2. when connected, look at IP-address in `tcp_conn_local_get`
> 3. use this address in SIP via header and other headers for local IP
> 
> this should work if the device has multiple network interfaces.

Sounds very promising.

**Comment by cspiel1** [comment]
> I would also like to suggest to split up this patch into logical parts.

Ok. I will squash the commits and split them again into logical parts.

> 
> for SIP multi-homing it should be quite easy to do it for TCP/TLS:
> 
> 1. TCP connect to the SIP server
> 2. when connected, look at IP-address in `tcp_conn_local_get`
> 3. use this address in SIP via header and other headers for local IP
> 
> this should work if the device has multiple network interfaces.
> 
> The SDP part is not so easy, and should be handled separately.

This PR is more concerned to Server-less calls to an IP-address.
- It selects correct address family and local IP address for SDP.
- It sets the scope ID for destination address if it is an IPv6 link local address.

For registered UAs maybe we could use the local IP of the registration.

**Comment by cspiel1** [comment]
@alfredh @juha-h @sreimers Would it be ok to change the API in order to set the `laddr` after the DNS resolution was done?
e.g.
- replace the `struct mbuf *desc` by a `struct struct sdp_session *desc` in function `sipsess_connect()`

**Comment by juha-h** [comment]
Christian Spielberger writes:

> @alfredh @juha-h @sreimers Would it be ok to change the API in order
> to set the `laddr` after the DNS resolution was done? 
> e.g.
> - replace the `struct mbuf *desc` by a `struct struct sdp_session
> *desc` in function `sipsess_connect()`

Should be OK, but in theory also SUBSCRIBE requests can contain bodies.

**Comment by cspiel1** [comment]
Split the PR into logical separated commits.

So far not added the suggested features discussed above.
- Select laddr after DNS resolution.
- Tcp connect before select laddr.

We should handle these in future PRs.

**Comment by alfredh** [comment]
is there an easy way to test IPv6 link-local on the local machine ?

**Comment by alfredh** [comment]
should we also support IPv6 address suffix?

```
fe80::b059:65f4:e877:c40%en0
```

https://superuser.com/questions/99746/why-is-there-a-percent-sign-in-the-ipv6-address

**Comment by cspiel1** [comment]
> is there an easy way to test IPv6 link-local on the local machine ?

I test always with two machines in local network. Both with baresip.
```
echo "/dial [fe80::...]" | nc -N localhost 5555
```

**Comment by cspiel1** [comment]
> should we also support IPv6 address suffix?
> 
> ```
> fe80::b059:65f4:e877:c40%en0
> ```
> 
> https://superuser.com/questions/99746/why-is-there-a-percent-sign-in-the-ipv6-address

This patch automatically selects right scope ID which is assigned to the interface. There is no need that the user specifies the network interface for now.

Anyhow it could make sense to support the address suffix in `sa_decode() or sa_set(), sa_set_str()`. We could postpone this until someone needs this.

**Comment by alfredh** [comment]
regarding multiple addresses and IPv6-linklocal;

it looks like this patch can be split into 2 parts:

1. add support for multiple SIP listen interfaces
2. add support for IPv6 link local


regarding setting the scope-id on struct sa:

if you imagine the lifetime of a `struct sa` object, it is not very clean to
change the scopeid during the middle of the sa lifetime. I think it is more
clear to set the scopeid either:

1. when the sa is inited
2. right before the sa is used to send a packet out on a udp/tcp socket

is this possible to do ?

**Comment by cspiel1** [comment]
- The split should be possible.
- You gave me an idea. When I slightly change `transp_find()` such that `dst` is not a const, it could "find" and set also the correct scope ID if `dst` is an IPv6ll.

**Comment by cspiel1** [comment]
Now removed the scope ID stuff for IPv6ll.

The PR contains now only the commits to support all local network addresses and interfaces.

IPv6ll is already listed but does not work. Setting the scope ID will be re-based on top of this.

**Comment by sreimers** [comment]
Makefile:148: *** bad libre version, required version is ">= 2.0.1-dev2". LIBRE_MK: /usr/local/share/re/re.mk.  Stop.

The libre PR needs a rebase, after this I can trigger a re-build.

**Comment by cspiel1** [comment]
> `Makefile:148: *** bad libre version, required version is ">= 2.0.1-dev2". LIBRE_MK: /usr/local/share/re/re.mk. Stop.`
> 
> The libre PR needs a rebase, after this I can trigger a re-build.

Did the rebase already, this triggered the re-build. But the error is new:
```
Err:7 https://packages.microsoft.com/ubuntu/20.04/prod focal Release
14
  Could not handshake: Error in the pull function. [IP: 40.74.238.15 443]
```

**Comment by sreimers** [comment]
triggered a re-build, seems to be fixed now.

**Comment by juha-h** [comment]
Should this test https://github.com/baresip/baresip/pull/1473#issuecomment-857494756 now work with this PR or does it only handle ipv6ll case?

**Comment by cspiel1** [comment]
> Should this test [#1473 (comment)](https://github.com/baresip/baresip/pull/1473#issuecomment-857494756) now work with this PR or does it only handle ipv6ll case?

Both no.
- DNS resolution before selecting laddr in ua.c would be another PR. But this is a first step.
- IPv6ll support will be added in https://github.com/baresip/baresip/pull/1500, which also depends on this PR.

**Comment by alfredh** [comment]
please take a look at this test-program:

```c

#include <string.h>
#include <re.h>
#include <net/if.h>
#define __USE_POSIX 1  /**< Use POSIX flag */
#define __USE_XOPEN2K 1/**< Use POSIX.1:2001 code */
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>


static int resolve_address(struct sa *sa, const char *addr)
{
	struct addrinfo *res, *res0;
	int error;

	error = getaddrinfo(addr, NULL, NULL, &res0);
	if (error != 0) {
		fprintf(stderr, "error in getaddrinfo: %s\n",
			gai_strerror(error));
		return ENOENT;
	}

	for (res = res0; res; res = res->ai_next) {

		char ifname[64];

		sa_set_sa(sa, res->ai_addr);

		re_printf("scopeid: %u (%s)\n",
			  sa_scopeid(sa),
			  if_indextoname(sa_scopeid(sa), ifname));
		break;
	}

	return 0;
}


static void tcp_estab_handler(void *arg)
{
	re_printf("estab\n");
}


static void tcp_recv_handler(struct mbuf *mb, void *arg)
{
	re_printf("recv: %b\n", mbuf_buf(mb), mbuf_get_left(mb));
}


static void tcp_close_handler(int err, void *arg)
{
	re_printf("closed\n");
}


int main(int argc, char *argv[])
{
	struct tcp_conn *tc;
	struct sa srv;
	int err;

	if (argc < 2)
		return 2;

	err = resolve_address(&srv, argv[1]);
	if (err)
		return err;

	sa_set_port(&srv, 22);

	re_printf("connecting to: %J\n", &srv);

	err = tcp_connect(&tc, &srv, tcp_estab_handler,
			  tcp_recv_handler, tcp_close_handler, NULL);
	if (err) {
		re_printf("connect error (%m)\n", err);
		return err;
	}

	re_main(NULL);
}
```

when running the program locally on OSX:

```
$ ./test fe80::96c6:91ff:fead:cdc9%en0
scopeid: 5 (en0)
connecting to: [fe80::96c6:91ff:fead:cdc9]:22
estab
recv: SSH-2.0-OpenSSH_7.9p1 Debian-10+deb10u2

^C
```

`getaddrinfo` is able to resolve the IPv6 link local address and also the Scope-ID suffix.

perhaps we should try to get the HTTP client working first with this format?
example:

```
http://fe80::96c6:91ff:fead:cdc9%en0/index.html
```

**Comment by cspiel1** [comment]
Yes, I know that `getaddrinfo` is able to decode the interface suffix. And `tcp_conn_alloc` already uses `getaddrinfo`. When trying  the HTTP client, I found that `sa_set_str` seems to deny the IPv6ll address with the suffix. It uses `net_inet_pton`. I will try to replace this by a `getaddrinfo` solution. This would be a different PR.

Note: This PR adds support for all local addresses which are collected by `getifaddr` which already sets the scope ID correctly. The goal should be that the user has not to specify the network interface for a IPv6ll target address. Instead it should be found by libre.

Anyway, I agree with you that the suffix should be supported. I will open another PR for this, if it works.

**Comment by alfredh** [comment]
after reading more about IPv6 linklocal and testing the suffix solution,
I believe we are in a better shape to take better technical decisions.

I would like to propose the following:

1. we add support for IPv6 link local address suffix in HTTP
2. we add support for IPv6 link local address suffix in SIP
3. we add support for multiple interfaces in SIP


step 3 will require changes in re and baresip at the same time.
we should not add support for multiple addresses in baresip
before it is working properly in re.


this PR is getting rather large. could you please squash all the commits into
one single commit and then make a new PR for it ?

**Comment by cspiel1** [comment]
will squash and create new PR.

File: include/baresip.h (hunk 3 above)
Comment (cspiel1) [reply include/baresip.h:692]: I tried to add the interface name to `struct laddr`. But this enlarges the PR. And the benefit is low.

**Comment by cspiel1** [comment]
> after reading more about IPv6 linklocal and testing the suffix solution,
> I believe we are in a better shape to take better technical decisions.
> 
> I would like to propose the following:
> 
> 1. we add support for IPv6 link local address suffix in HTTP
https://github.com/baresip/re/pull/116 is enough for this. No baresip change needed for HTTP.
> 2. we add support for IPv6 link local address suffix in SIP
In re, nothing in addition to the re PR is needed. As a wrote already, we need the correct link local address in baresip. So we need therefore already PR https://github.com/baresip/baresip/pull/1513.
> 3. we add support for multiple interfaces in SIP

**Pull request closed without merging**